File: web-api/providers/huawei.mjs
import crypto from "node:crypto";
import { writeApiLog } from "../core/database.mjs";
import { decryptSecret } from "../core/crypto.mjs";
import { getAccountRegion, getAccountSecretRecord } from "../repositories/accounts.mjs";

function encode(value) {
  return encodeURIComponent(String(value)).replace(/[!'()*]/g, (c) => `%${c.charCodeAt(0).toString(16).toUpperCase()}`);
}

function canonicalUri(pathname) {
  return pathname.split("/").map((part) => encode(decodeURIComponent(part))).join("/") || "/";
}

function queryText(query = {}) {
  return Object.entries(query).filter(([, value]) => value !== undefined && value !== null && value !== "")
    .map(([key, value]) => [encode(key), encode(value)])
    .sort(([leftKey, leftValue], [rightKey, rightValue]) => leftKey.localeCompare(rightKey) || leftValue.localeCompare(rightValue))
    .map(([key, value]) => `${key}=${value}`).join("&");
}

function errorMessage(data, status) {
  return data?.error_msg || data?.message || data?.error?.message || data?.code || `华为云 ${status}`;
}

export async function request(accountId, host, pathname, query = {}) {
  const row = getAccountSecretRecord(accountId);
  if (!row) throw new Error("云账号不存在");
  if (!row.enabled) throw new Error("云账号已停用");
  if (row.cloud_type !== "huawei") throw new Error("当前账号不是华为云账号");
  const encodedQuery = queryText(query);
  const date = new Date().toISOString().replace(/[-:]|\.\d{3}/g, "");
  const headers = { host, "x-sdk-date": date };
  const canonicalHeaders = Object.entries(headers).map(([key, value]) => `${key}:${value}\n`).join("");
  const signedHeaders = Object.keys(headers).join(";");
  const payloadHash = crypto.createHash("sha256").update("").digest("hex");
  const canonicalRequest = `GET\n${canonicalUri(pathname)}\n${encodedQuery}\n${canonicalHeaders}\n${signedHeaders}\n${payloadHash}`;
  const stringToSign = `SDK-HMAC-SHA256\n${date}\n${crypto.createHash("sha256").update(canonicalRequest).digest("hex")}`;
  const signature = crypto.createHmac("sha256", decryptSecret(row.secret_ciphertext)).update(stringToSign).digest("hex");
  const authorization = `SDK-HMAC-SHA256 Access=${row.access_key_id}, SignedHeaders=${signedHeaders}, Signature=${signature}`;
  const response = await fetch(`https://${host}${canonicalUri(pathname)}${encodedQuery ? `?${encodedQuery}` : ""}`, { headers: { Host: host, "X-Sdk-Date": date, Authorization: authorization } });
  const text = await response.text();
  let data;
  try { data = JSON.parse(text); } catch { data = { message: text || `华为云 ${response.status}` }; }
  if (!response.ok) {
    const message = errorMessage(data, response.status);
    writeApiLog(accountId, host, `GET ${pathname}`, query, data, "失败", message);
    throw new Error(message);
  }
  writeApiLog(accountId, host, `GET ${pathname}`, query, data, "成功");
  return data;
}

export async function context(accountId) {
  const account = getAccountRegion(accountId);
  const defaultRegion = String(account?.region_id || "cn-north-4");
  const data = await request(accountId, `iam.${defaultRegion}.myhuaweicloud.com`, "/v3/projects", { enabled: "true" });
  const projects = Array.isArray(data?.projects) ? data.projects.filter((item) => item.id && item.name && String(item.status || "enabled").toLowerCase() === "enabled") : [];
  if (!projects.length) throw new Error("未读取到可用项目，请检查 IAM 项目权限");
  return { defaultRegion, projects };
}

function arrayAt(value, path) { let current = value; for (const key of path) current = current?.[key]; return Array.isArray(current) ? current : []; }
async function offsetPages(fetchPage, path, pageSize = 100) {
  const items = [];
  for (let offset = 0; offset < 10_000; offset += pageSize) { const page = arrayAt(await fetchPage(offset), path); items.push(...page); if (page.length < pageSize) return items; }
  throw new Error("分页超过 100 页，已停止读取");
}
function instance(item, region, project) { const addresses = Object.values(item.addresses || {}).flat(); const publicIp = addresses.find((address) => String(address?.["OS-EXT-IPS:type"] || "").toLowerCase() === "floating")?.addr || ""; const privateIp = addresses.find((address) => String(address?.["OS-EXT-IPS:type"] || "").toLowerCase() !== "floating")?.addr || ""; return { ...item, InstanceId: item.id, InstanceName: item.name || item.id, InstanceStatus: item.status, Status: item.status, PublicIpAddress: publicIp, PrivateIpAddress: privateIp, InstanceType: item.flavor?.id || item.flavor?.name || "", VpcId: item.metadata?.vpc_id || "", _region_id: region, _project_id: project.id }; }
function rds(item, region, project) { return { ...item, DBInstanceId: item.id, DBInstanceDescription: item.name || item.id, DBInstanceStatus: item.status, DBInstanceClass: item.flavor_ref || item.flavor?.id || "", DBInstanceStorage: Number(item.volume?.size || 0), ConnectionString: item.private_ips?.[0] || item.nodes?.[0]?.private_ip || "", Port: item.port || "", Engine: item.datastore?.type || "", EngineVersion: item.datastore?.version || "", CreateTime: item.created || "", _region_id: region, _project_id: project.id }; }
function redis(item, region, project) { return { ...item, InstanceId: item.instance_id || item.id, InstanceName: item.name || item.instance_id || item.id, InstanceStatus: item.status || item.operating_status, InstanceType: item.engine || "Redis", InstanceClass: item.specification || item.capacity || "", Capacity: Number(item.capacity || 0) * 1024, ConnectionDomain: item.ip || item.private_ip || "", Port: item.port || "", EngineVersion: item.engine_version || "", NetworkType: item.vpc_name || "", _region_id: region, _project_id: project.id }; }
function zone(item) { return { ...item, DomainName: String(item.name || "").replace(/\.$/, ""), DomainStatus: item.status || "ACTIVE", ZoneId: item.id, RecordCount: Number(item.record_num || 0), RegistrationDate: item.created_at || "", _region_id: "cn-north-4", _huawei_public_zone: true }; }
function xmlText(xml, tag) { const match = String(xml).match(new RegExp(`<${tag}(?:\\s[^>]*)?>([\\s\\S]*?)</${tag}>`, "i")); return match ? match[1].trim() : ""; }
function xmlBlocks(xml, tag) { return [...String(xml).matchAll(new RegExp(`<${tag}(?:\\s[^>]*)?>([\\s\\S]*?)</${tag}>`, "gi"))].map((match) => match[1]); }

async function obsBuckets(accountId, region) {
  const row = getAccountSecretRecord(accountId);
  if (!row || !row.enabled || row.cloud_type !== "huawei") throw new Error("当前账号不是华为云账号");
  const host = `obs.${region}.myhuaweicloud.com`;
  const date = new Date().toUTCString();
  const signature = crypto.createHmac("sha1", decryptSecret(row.secret_ciphertext)).update(`GET\n\n\n${date}\n/`).digest("base64");
  const response = await fetch(`https://${host}/`, { headers: { Date: date, Host: host, Authorization: `OBS ${row.access_key_id}:${signature}` } });
  const xml = await response.text();
  if (!response.ok) { const message = `OBS ${response.status}: ${xmlText(xml, "Message") || xmlText(xml, "Code") || "请求被拒绝"}`; writeApiLog(accountId, host, "ListBuckets", {}, { body: xml }, "失败", message); throw new Error(message); }
  const buckets = xmlBlocks(xml, "Bucket").map((block) => ({ Name: xmlText(block, "Name"), BucketName: xmlText(block, "Name"), Location: xmlText(block, "Location") || region, CreationDate: xmlText(block, "CreationDate"), StorageClass: "STANDARD", Acl: "private", ExtranetEndpoint: "-", IntranetEndpoint: "-", _region_id: xmlText(block, "Location") || region })).filter((bucket) => bucket.Name);
  writeApiLog(accountId, host, "ListBuckets", {}, { count: buckets.length }, "成功");
  return buckets;
}

export async function resources(accountId, type) {
  const { defaultRegion, projects } = await context(accountId);
  const items = [];
  const errors = [];
  if (type === "domain") { try { items.push(...await offsetPages((offset) => request(accountId, "dns.cn-north-4.myhuaweicloud.com", "/v2/zones", { limit: 500, offset }), ["zones"], 500)); } catch (error) { errors.push(`cn-north-4: ${error.message}`); } return { resource_type: type, items: items.map(zone), errors, fetched_at: Date.now() }; }
  if (type === "oss") { for (const region of [...new Set([defaultRegion, ...projects.map((project) => project.name)])]) { try { items.push(...await obsBuckets(accountId, region)); } catch (error) { errors.push(`${region}: ${error.message}`); } } return { resource_type: type, items: Array.from(new Map(items.map((item) => [item.Name, item])).values()), errors, fetched_at: Date.now() }; }
  const definitions = { ecs: ["ecs", "/v1/{project}/cloudservers/detail", "servers", instance], rds: ["rds", "/v3/{project}/instances", "instances", rds], redis: ["dcs", "/v2/{project}/instances", "instances", redis] };
  const definition = definitions[type];
  if (!definition) return { resource_type: type, items, errors: [`华为云暂未接入 ${type} 资源`], fetched_at: Date.now() };
  const [service, template, path, normalize] = definition;
  for (const project of projects) { const region = project.name; try { const values = await offsetPages((offset) => request(accountId, `${service}.${region}.myhuaweicloud.com`, template.replace("{project}", encodeURIComponent(project.id)), { limit: 100, offset }), [path], 100); items.push(...values.map((item) => normalize(item, region, project))); } catch (error) { errors.push(`${region}: ${error.message}`); } }
  return { resource_type: type, items, errors, fetched_at: Date.now() };
}

export async function verify(accountId) {
  const { defaultRegion, projects } = await context(accountId);
  const regions = [...new Set(projects.map((project) => project.name))];
  return { provider: "huawei", verified: true, region_count: regions.length, regions, default_region: defaultRegion, project_count: projects.length };
}

File: src-tauri/src/cloud/huawei.rs
use crate::{
    account_cloud_type, account_credentials, array_at, rpc_encode, string_params,
    write_api_log, ResourceResponse,
};
use crate::core::storage::open_db;
use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use chrono::Utc;
use hmac::{Hmac, Mac};
use serde_json::{json, Value};
use sha1::Sha1;
use sha2::{Digest, Sha256};
use std::{collections::BTreeMap, future::Future};

fn encode(value: &str) -> String { rpc_encode(value) }

fn canonical_uri(path: &str) -> String {
    let value = path.split('/').map(encode).collect::<Vec<_>>().join("/");
    if value.is_empty() { "/".into() } else { value }
}

fn query_string(query: &BTreeMap<String, String>) -> String {
    let mut values = query.iter().filter(|(_, value)| !value.is_empty()).map(|(key, value)| (encode(key), encode(value))).collect::<Vec<_>>();
    values.sort();
    values.into_iter().map(|(key, value)| format!("{key}={value}")).collect::<Vec<_>>().join("&")
}

fn error_message(data: &Value, status: reqwest::StatusCode) -> String {
    data.get("error_msg").or_else(|| data.get("message")).or_else(|| data.pointer("/error/message")).or_else(|| data.get("code")).and_then(Value::as_str).unwrap_or(&format!("华为云 {status}")).to_string()
}

async fn request(id: i64, host: &str, path: &str, query: BTreeMap<String, String>) -> Result<Value, String> {
    if account_cloud_type(id)? != "huawei" { return Err("当前账号不是华为云账号".into()); }
    let (access_key_id, access_key_secret) = account_credentials(id)?;
    let query_text = query_string(&query);
    let canonical_uri = canonical_uri(path);
    let date = Utc::now().format("%Y%m%dT%H%M%SZ").to_string();
    let canonical_headers = format!("host:{host}\nx-sdk-date:{date}\n");
    let payload_hash = format!("{:x}", Sha256::digest(b""));
    let canonical_request = format!("GET\n{canonical_uri}\n{query_text}\n{canonical_headers}\nhost;x-sdk-date\n{payload_hash}");
    let string_to_sign = format!("SDK-HMAC-SHA256\n{date}\n{:x}", Sha256::digest(canonical_request.as_bytes()));
    let mut mac: Hmac<Sha256> = <Hmac<Sha256> as Mac>::new_from_slice(access_key_secret.as_bytes()).map_err(|error| error.to_string())?;
    mac.update(string_to_sign.as_bytes());
    let authorization = format!("SDK-HMAC-SHA256 Access={access_key_id}, SignedHeaders=host;x-sdk-date, Signature={}", hex::encode(mac.finalize().into_bytes()));
    let url = format!("https://{host}{canonical_uri}{}", if query_text.is_empty() { String::new() } else { format!("?{query_text}") });
    let response = reqwest::Client::new().get(url).header("Host", host).header("X-Sdk-Date", &date).header("Authorization", authorization).timeout(std::time::Duration::from_secs(30)).send().await.map_err(|error| format!("华为云请求失败: {error}"))?;
    let status = response.status();
    let text = response.text().await.map_err(|error| format!("华为云返回读取失败: {error}"))?;
    let data: Value = serde_json::from_str(&text).unwrap_or_else(|_| json!({"message": text}));
    if !status.is_success() {
        let message = error_message(&data, status);
        write_api_log(&access_key_id, host, &format!("GET {path}"), &json!(query), Some(&data), "失败", Some(&message));
        return Err(message);
    }
    write_api_log(&access_key_id, host, &format!("GET {path}"), &json!(query), Some(&data), "成功", None);
    Ok(data)
}

async fn offset_pages<F, Fut>(mut fetch_page: F, path: &[&str], page_size: usize) -> Result<Vec<Value>, String>
where
    F: FnMut(usize) -> Fut,
    Fut: Future<Output = Result<Value, String>>,
{
    let mut items = Vec::new();
    for offset in (0..10_000).step_by(page_size) {
        let data = fetch_page(offset).await?;
        let page = array_at(&data, path).into_iter().cloned().collect::<Vec<_>>();
        let count = page.len();
        items.extend(page);
        if count < page_size { return Ok(items); }
    }
    Err("分页超过 100 页，已停止读取".into())
}

async fn context(id: i64) -> Result<(String, Vec<Value>), String> {
    let default_region = open_db()?.query_row("SELECT region_id FROM cloud_accounts WHERE id=?1", [id], |row| row.get::<_, Option<String>>(0)).map_err(|error| error.to_string())?.filter(|value| !value.trim().is_empty()).unwrap_or_else(|| "cn-north-4".into());
    let data = request(id, &format!("iam.{default_region}.myhuaweicloud.com"), "/v3/projects", string_params(&[("enabled", "true".into())])).await?;
    let projects = array_at(&data, &["projects"]).into_iter().filter(|project| project.get("id").and_then(Value::as_str).is_some_and(|value| !value.is_empty()) && project.get("name").and_then(Value::as_str).is_some_and(|value| !value.is_empty()) && project.get("status").and_then(Value::as_str).map(|value| value.eq_ignore_ascii_case("enabled")).unwrap_or(true)).cloned().collect::<Vec<_>>();
    if projects.is_empty() { return Err("未读取到可用项目，请检查 IAM 项目权限".into()); }
    Ok((default_region, projects))
}

fn instance(item: &Value, region: &str, project: &Value) -> Value {
    let addresses = item.get("addresses").and_then(Value::as_object).into_iter().flat_map(|group| group.values()).filter_map(Value::as_array).flatten().collect::<Vec<_>>();
    let public_ip = addresses.iter().find(|address| address.get("OS-EXT-IPS:type").and_then(Value::as_str).is_some_and(|kind| kind.eq_ignore_ascii_case("floating"))).and_then(|address| address.get("addr")).cloned().unwrap_or(json!(""));
    let private_ip = addresses.iter().find(|address| !address.get("OS-EXT-IPS:type").and_then(Value::as_str).is_some_and(|kind| kind.eq_ignore_ascii_case("floating"))).and_then(|address| address.get("addr")).cloned().unwrap_or(json!(""));
    let mut value = item.clone();
    if let Value::Object(ref mut target) = value {
        target.insert("InstanceId".into(), item.get("id").cloned().unwrap_or(json!(""))); target.insert("InstanceName".into(), item.get("name").or_else(|| item.get("id")).cloned().unwrap_or(json!("")));
        target.insert("InstanceStatus".into(), item.get("status").cloned().unwrap_or(json!(""))); target.insert("Status".into(), item.get("status").cloned().unwrap_or(json!("")));
        target.insert("PublicIpAddress".into(), public_ip); target.insert("PrivateIpAddress".into(), private_ip); target.insert("InstanceType".into(), item.pointer("/flavor/id").or_else(|| item.pointer("/flavor/name")).cloned().unwrap_or(json!("")));
        target.insert("VpcId".into(), item.pointer("/metadata/vpc_id").cloned().unwrap_or(json!(""))); target.insert("_region_id".into(), json!(region)); target.insert("_project_id".into(), project.get("id").cloned().unwrap_or(json!("")));
    }
    value
}

fn rds(item: &Value, region: &str, project: &Value) -> Value {
    json!({"DBInstanceId": item.get("id"), "DBInstanceDescription": item.get("name").or_else(|| item.get("id")), "DBInstanceStatus": item.get("status"), "DBInstanceClass": item.get("flavor_ref").or_else(|| item.pointer("/flavor/id")), "DBInstanceStorage": item.pointer("/volume/size").unwrap_or(&json!(0)), "ConnectionString": item.pointer("/private_ips/0").or_else(|| item.pointer("/nodes/0/private_ip")), "Port": item.get("port"), "Engine": item.pointer("/datastore/type"), "EngineVersion": item.pointer("/datastore/version"), "CreateTime": item.get("created"), "_region_id": region, "_project_id": project.get("id"), "_raw": item})
}

fn redis(item: &Value, region: &str, project: &Value) -> Value {
    let capacity = item.get("capacity").and_then(|value| value.as_f64().or_else(|| value.as_str()?.parse().ok())).unwrap_or(0.0) * 1024.0;
    json!({"InstanceId": item.get("instance_id").or_else(|| item.get("id")), "InstanceName": item.get("name").or_else(|| item.get("instance_id")).or_else(|| item.get("id")), "InstanceStatus": item.get("status").or_else(|| item.get("operating_status")), "InstanceType": item.get("engine").unwrap_or(&json!("Redis")), "InstanceClass": item.get("specification").or_else(|| item.get("capacity")), "Capacity": capacity, "ConnectionDomain": item.get("ip").or_else(|| item.get("private_ip")), "Port": item.get("port"), "EngineVersion": item.get("engine_version"), "NetworkType": item.get("vpc_name"), "_region_id": region, "_project_id": project.get("id"), "_raw": item})
}

fn zone(item: &Value) -> Value {
    let name = item.get("name").and_then(Value::as_str).unwrap_or("").trim_end_matches('.');
    json!({"DomainName": name, "DomainStatus": item.get("status").unwrap_or(&json!("ACTIVE")), "ZoneId": item.get("id"), "RecordCount": item.get("record_num").unwrap_or(&json!(0)), "RegistrationDate": item.get("created_at"), "_region_id": "cn-north-4", "_huawei_public_zone": true, "_raw": item})
}

fn xml_text(body: &str, tag: &str) -> String {
    let open = format!("<{tag}>"); let close = format!("</{tag}>");
    body.find(&open).and_then(|start| body[start + open.len()..].find(&close).map(|end| body[start + open.len()..start + open.len() + end].to_string())).unwrap_or_default()
}

fn xml_blocks(body: &str, tag: &str) -> Vec<String> {
    let open = format!("<{tag}>"); let close = format!("</{tag}>"); let mut values = Vec::new(); let mut rest = body;
    while let Some(start) = rest.find(&open) { let chunk = &rest[start + open.len()..]; let Some(end) = chunk.find(&close) else { break }; values.push(chunk[..end].to_string()); rest = &chunk[end + close.len()..]; }
    values
}

async fn obs_buckets(id: i64, region: &str) -> Result<Vec<Value>, String> {
    if account_cloud_type(id)? != "huawei" { return Err("当前账号不是华为云账号".into()); }
    let (access_key_id, access_key_secret) = account_credentials(id)?;
    let host = format!("obs.{region}.myhuaweicloud.com");
    let date = Utc::now().format("%a, %d %b %Y %H:%M:%S GMT").to_string();
    let mut mac: Hmac<Sha1> = <Hmac<Sha1> as Mac>::new_from_slice(access_key_secret.as_bytes()).map_err(|error| error.to_string())?;
    mac.update(format!("GET\n\n\n{date}\n/").as_bytes());
    let response = reqwest::Client::new().get(format!("https://{host}/")).header("Date", &date).header("Host", &host).header("Authorization", format!("OBS {access_key_id}:{}", B64.encode(mac.finalize().into_bytes()))).timeout(std::time::Duration::from_secs(30)).send().await.map_err(|error| format!("OBS 请求失败: {error}"))?;
    let status = response.status();
    let body = response.text().await.map_err(|error| error.to_string())?;
    if !status.is_success() {
        let message = format!("OBS {status}: {}", { let value = xml_text(&body, "Message"); if value.is_empty() { xml_text(&body, "Code") } else { value } });
        write_api_log(&access_key_id, &host, "ListBuckets", &json!({}), Some(&json!({"body": body})), "失败", Some(&message));
        return Err(message);
    }
    let buckets = xml_blocks(&body, "Bucket").into_iter().map(|bucket| {
        let name = xml_text(&bucket, "Name");
        let location = { let value = xml_text(&bucket, "Location"); if value.is_empty() { region.to_string() } else { value } };
        json!({"Name": name, "BucketName": name, "Location": location, "CreationDate": xml_text(&bucket, "CreationDate"), "StorageClass": "STANDARD", "Acl": "private", "ExtranetEndpoint": "-", "IntranetEndpoint": "-", "_region_id": location})
    }).filter(|bucket| bucket.get("Name").and_then(Value::as_str).is_some_and(|value| !value.is_empty())).collect::<Vec<_>>();
    write_api_log(&access_key_id, &host, "ListBuckets", &json!({}), Some(&json!({"count": buckets.len()})), "成功", None);
    Ok(buckets)
}

pub(crate) async fn resource_items(id: i64, resource_type: &str) -> ResourceResponse {
    let now = Utc::now().timestamp_millis();
    let (default_region, projects) = match context(id).await { Ok(value) => value, Err(error) => return ResourceResponse { resource_type: resource_type.into(), items: vec![], errors: vec![error], fetched_at: now } };
    let mut items = Vec::new(); let mut errors = Vec::new();
    if resource_type == "domain" {
        match offset_pages(|offset| request(id, "dns.cn-north-4.myhuaweicloud.com", "/v2/zones", string_params(&[("limit", "500".into()), ("offset", offset.to_string())])), &["zones"], 500).await { Ok(values) => items.extend(values.into_iter().map(|item| zone(&item))), Err(error) => errors.push(format!("cn-north-4: {error}")), }
        return ResourceResponse { resource_type: resource_type.into(), items, errors, fetched_at: now };
    }
    if resource_type == "oss" {
        let mut regions = projects.iter().filter_map(|project| project.get("name").and_then(Value::as_str)).map(String::from).collect::<Vec<_>>(); regions.push(default_region); regions.sort(); regions.dedup();
        for region in regions { match obs_buckets(id, &region).await { Ok(values) => items.extend(values), Err(error) => errors.push(format!("{region}: {error}")), } }
        let mut unique = BTreeMap::new(); for item in items { if let Some(name) = item.get("Name").and_then(Value::as_str) { unique.insert(name.to_string(), item); } }
        return ResourceResponse { resource_type: resource_type.into(), items: unique.into_values().collect(), errors, fetched_at: now };
    }
    let service = match resource_type { "ecs" => Some(("ecs", "/v1/{project}/cloudservers/detail", "servers")), "rds" => Some(("rds", "/v3/{project}/instances", "instances")), "redis" => Some(("dcs", "/v2/{project}/instances", "instances")), _ => None };
    let Some((service_name, path_template, response_path)) = service else { return ResourceResponse { resource_type: resource_type.into(), items, errors: vec![format!("华为云暂未接入 {resource_type} 资源")], fetched_at: now }; };
    for project in &projects {
        let region = project.get("name").and_then(Value::as_str).unwrap_or(""); let project_id = project.get("id").and_then(Value::as_str).unwrap_or("");
        if region.is_empty() || project_id.is_empty() { continue; }
        let path = path_template.replace("{project}", &encode(project_id)); let host = format!("{service_name}.{region}.myhuaweicloud.com");
        match offset_pages(|offset| request(id, &host, &path, string_params(&[("limit", "100".into()), ("offset", offset.to_string())])), &[response_path], 100).await {
            Ok(values) => for item in values { items.push(match resource_type { "ecs" => instance(&item, region, project), "rds" => rds(&item, region, project), "redis" => redis(&item, region, project), _ => item }); },
            Err(error) => errors.push(format!("{region}: {error}")),
        }
    }
    ResourceResponse { resource_type: resource_type.into(), items, errors, fetched_at: now }
}

#[tauri::command]
pub(crate) async fn verify_huawei_account(id: i64) -> Result<Value, String> {
    let (default_region, projects) = context(id).await?;
    let mut regions = projects.iter().filter_map(|project| project.get("name").and_then(Value::as_str)).map(String::from).collect::<Vec<_>>(); regions.sort(); regions.dedup();
    Ok(json!({"provider": "huawei", "verified": true, "region_count": regions.len(), "regions": regions, "default_region": default_region, "project_count": projects.len()}))
}
